### ISSUE TITLE
Overriding WordPress globals is prohibited.

### ISSUE THREAD

**Issue by iruzevic** [WordPress-Coding-Standards/WordPress-Coding-Standards#1490]
phpcs: 3.3.1
wpcs: 3.3.1
php: 7.1.16

I'm creating a plugin but the same thing happens if I'm creating a theme.

**Steps:**
I'm creating a new page:
```php
add_menu_page(
      esc_html__( 'Test', 'test' ),
      esc_html__( 'Test', 'test' ),
      'edit_posts'
      'test'
      array( $this, 'get_welcome_page' ),
      'dashicons-laptop'
);
```

providing template callback:
```php
  public function get_welcome_page() {
    require_once plugin_dir_path( dirname( __FILE__ ) ) . 'admin/settings/welcome.php';
  }
```

in template I'm assing some value to $title, $id, $menu, $paged, $post_id variable:
```php
$title = 'some value';
```

and i get linting error:
```
Overriding WordPress globals is prohibited. Found assignment to $title
    |       | (WordPress.WP.GlobalVariablesOverride.OverrideProhibited)
```

I think that this is a false positive because I'm not using any global variable inside this tempalte and I would like to use $title variable.

**Comment by GaryJones** [comment]
On its own, the `$title` in the view file appears to be in the global scope, which is why it is flagged. It's only at run-time do we see that the view file is required inside of a function.

PHPCS doesn't execute the PHP, so it can't detect this.

You've got a few main options:
 - Don't set the `$title` in the view file, but set it inside `get_welcome_page()`, and then just escape and echo it in the view file (keep the logic out of the view file).
 - Echo `'some value'` directly where you need it and don't use the variable at all.
 - Use `// phpcs:ignore WordPress.WP.GlobalVariablesOverride.OverrideProhibited` in the line before the assignment to `$title` in the view file.
 - Choose a different but equally as useful variable name.

**Comment by jrfnl** [comment]
What @GaryJones said + it looks like you're not using the `PrefixAllGlobals` sniff as that would have complained as well.

The problem with your code is that, while in your example case, you are including the view file from within a function call, making the variable a local variable, there is no **guarantee** that the file will _always_ be included from within a function call - think other plugins or themes interacting with yours, but also hackers -.

So, I would like to strongly suggest to enable the `PrefixAllGlobals` sniff and to prefix the `$title` variable with your project prefix.

See this page in the wiki about how to enable the `PrefixAllGlobals` sniff: https://github.com/WordPress-Coding-Standards/WordPress-Coding-Standards/wiki/Customizable-sniff-properties#naming-conventions-prefix-everything-in-the-global-namespace

**Comment by iruzevic** [comment]
I understand and get all your points but:

@GaryJones your options are valid but let's say that I'm using namespacing in a plugin so everything that I define inside that namespace will live only there. Shiff should be able to se if I'm using namespacing or not. Also asking developer not to use for example $title variable because WP uses it not so friendly towards the developers. Also, I can't just echo the variable that I got from somewhere else because I can't know that that variable will exist or not.

@GaryJones This is a completely valid point if I'm writing on a theme that lives inside global namespace. But I'm using namespacing inside my plugin and I think that I'm safe with globals overriding my code.
Please correct me if I'm wrong.

**Comment by jrfnl** [comment]
> let's say that I'm using namespacing in a plugin so everything that I define inside that namespace will live only there. Shiff should be able to se if I'm using namespacing or not. 

1. The sniff does detect and take namespaces into account.
2. $variables are not namespaced, so what you say is incorrect and your code is still overriding the global variable.

**Comment by iruzevic** [comment]
ok tnx for your answers. I will close this issue.
